lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1507,6 +1509,59 @@ static void replaceOperandWithAnnotationIntrinsicCallResult(Value *&V) {
   }
 }
 
+// Translate aliasing memory access masks for SPIRVLoad and SPIRVStore
+// instructions. These masks are mapped on alias.scope and noalias
+// metadata in LLVM. Translation of optional string operand isn't yet supported
+// in the translator.
+template <typename SPIRVInstType>
+void SPIRVToLLVM::transAliasingMemAccess(SPIRVInstType *BI, Instruction *I) {
+  static_assert(std::is_same<SPIRVInstType, SPIRVStore>::value ||
+                std::is_same<SPIRVInstType, SPIRVLoad>::value,
+                "Only stores and loads can be aliased by memory access mask");
+  bool IsAliasScope = BI->SPIRVMemoryAccess::isAliasScope();
+  bool IsNoAlias = BI->SPIRVMemoryAccess::isNoAlias();
+  if (!(IsAliasScope || IsNoAlias))
+    return;
+  uint32_t AliasMDKind = IsAliasScope ? LLVMContext::MD_alias_scope
+                                      : LLVMContext::MD_noalias;
+  SPIRVId AliasListId = BI->SPIRVMemoryAccess::getAliasing();
+  addMemAliasMetadata(I, AliasListId, AliasMDKind);
+}
+
+// Create and apply alias.scope/noalias metadata
+void SPIRVToLLVM::addMemAliasMetadata(Instruction *I, SPIRVId AliasListId,
+                                      uint32_t AliasMDKind) {
+  SPIRVAliasScopeListDeclINTEL *AliasList =
+      BM->get<SPIRVAliasScopeListDeclINTEL>(AliasListId);
+  std::vector<SPIRVId> AliasScopeIds = AliasList->getArguments();
+  MDBuilder MDB(*Context);
+  SmallVector<Metadata *, 4> MDScopes;
+  for (const auto ScopeId : AliasScopeIds) {
+    SPIRVAliasScopeDeclINTEL *AliasScope =
+        BM->get<SPIRVAliasScopeDeclINTEL>(ScopeId);
+    std::vector<SPIRVId> AliasDomainIds = AliasScope->getArguments();
+    // Currently we expect exactly one argument for aliasing scope
+    // instruction.
+    // TODO: add translation of string scope and domain operand.
+    assert(AliasDomainIds.size() == 1 &&
+           "AliasScopeDeclINTEL must have exactly one argument");
+    SPIRVId AliasDomainId = AliasDomainIds[0];
+    // Create and store unique domain and scope metadata
+    MDAliasDomainMap.emplace(std::pair<SPIRVId, MDNode *>(
```

Comment:
Usage of `std::unordered_map` promises an OK performance, but it's not the fasted algorithm here. A bunch of if-elses would be a bit faster, but here I'd prefer code-readability, than some extra optimizations. WDYT?

Reply:
I don't think this is the case when we should care about performance.

> The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; premature optimization is the root of all evil (or at least most of it) in programming.
> -- Donald Knuth.

Speaking about readability, I guess, we can write it as:
```suggestion
    MDAliasDomainMap[AliasDomainId] = MDB.createAnonymousAliasScopeDomain();
```

Reply:
The suggestion won't work, since `emplace` checks if there is an element in map and add it only if it's not, and [] operator will insert/replace this element no matter what. So in the end it would be different aliasing scopes.

Reply:
Then maybe we can do 
```suggestion
    MDAliasDomainMap.emplace(AliasDomainId, MDB.createAnonymousAliasScopeDomain());
```
?